l10n_th_tin_service/models/res_partner.py

```diff
@@ -0,0 +1,196 @@
+           :param tin: a string for TIN or PIN
+           :param branch: one digit of branch number
+        """
+        branch = int(branch)
+        sess = Session()
+        mod_dir = path.dirname(path.realpath(__file__))
+        cert_path = str(Path(mod_dir).parents[0]) + "/static/cert/adhq1_ADHQ5.cer"
+        sess.verify = cert_path
+        transp = Transport(session=sess)
+        try:
+            cl = Client(ResPartner.vat_web_service_url, transport=transp)
+        except requests.exceptions.SSLError:
+            _logger.log(INFO, "Fall back to unverifed HTTPS request.")
+            sess.verify = False
+            transp = Transport(session=sess)
+            cl = Client(ResPartner.vat_web_service_url, transport=transp)
+        result = cl.service.Service(
+            "anonymous",
+            "anonymous",
+            TIN=tin,
+            ProvinceCode=0,
+            BranchNumber=branch,
+            AmphurCode=0,
+        )
+        odata = helpers.serialize_object(result)
+        _logger.log(DEBUG, pprint.pformat(odata))
+        data = OrderedDict()
+        if odata["vmsgerr"] is None:
+            for key, value in odata.items():
+                if (
+                    value is None
+                    or value["anyType"][0] == "-"
+                    or key in {"vNID", "vBusinessFirstDate"}
+                ):
+                    continue
+                data[key] = value["anyType"][0]
+            return data
+        return False
+
+    @api.onchange("vat", "branch")
+    def _onchange_vat_branch(self):
+        word_map = {
+            "vBuildingName": "อาคาร ",
+            "vFloorNumber": "ชั้นที่ ",
+            "vVillageName": "หมู่บ้าน ",
+            "vRoomNumber": "ห้องเลขที่ ",
+            "vHouseNumber": "เลขที่ ",
+            "vMooNumber": "หมู่ที่ ",
+            "vSoiName": "ซอย ",
+            "vStreetName": "ถนน ",
+            "vThambol": "ต.",
+            "vAmphur": "อ.",
+            "vProvince": "จ.",
+            "vPostCode": "",
+        }
+        street_map = [
+            [
+                "vBuildingName",
+                "vRoomNumber",
+                "vFloorNumber",
+                "vHouseNumber",
+                "vStreetName",
+                "vSoiName",
+            ],
+            ["vVillageName", "vMooNumber", "vThambol"],
+        ]
+        check_branch = re.compile(r"^\d{5}$")
+        if self.branch is False:
+            self.branch = "00000"
+
+        if self.vat:
+            if ResPartner.check_rd_tin_service(self.vat):
+                match = check_branch.match(self.branch)
+                if match is None:
+                    return {
+                        "warning": {
+                            "title": _("Branch validation failed"),
+                            "message": _(
+                                "Branch number %s must be 5 digits." % self.branch
+                            ),
+                        }
+                    }
```

Comment:
The branch validation should be implemented separately in a dedicated `@api.constrains('branch')` method.

```suggestion
    # put this outside of the _onchange_vat_branch() method
    @api.constrains('branch', 'country_id')
    def _check_branch(self):
        for record in self:
            if "TH" != record.country_id.code or not record.branch:
                continue
            check_branch = re.compile(r"^\d{5}$")
            match = check_branch.match(self.branch)
            if match is None:
                return {"warning": {
                            "title": _("Branch validation failed"),
                            "message": _("Branch number %s must be 5 digits." % self.branch), }
                        }            
```